src/completions.jl

```diff
@@ -101,6 +107,87 @@ completion(mod, c, prefix) = CompletionSuggetion(
   :detailtype         => completiondetailtype(c)
 )
 
+const MethodCompletionInfo = NamedTuple{(:f,:m,:tt),Tuple{Any,Method,Type}}
+const MethodCompletionDetail = NamedTuple{(:rt,:desc),Tuple{String,String}}
+"""
+    METHOD_COMPLETION_CACHE::Dict{String,Union{MethodCompletionInfo,MethodCompletionDetail}}
+    MethodCompletionInfo = NamedTuple{(:f,:m,:tt),Tuple{Any,Method,Type}}
+    MethodCompletionDetail = NamedTuple{(:rt,:desc),Tuple{String,String}}
+
+`Dict` that caches details of `REPLCompletions.MethodCompletion`s:
+- result of a return type inference
+- documentation description
+
+This cache object has the following structure:
+Keys are hash `String` of a `REPLCompletions.MethodCompletion` object,
+  which can be used for cache detection or refered lazily by
+  [autocompluete-plus's `getSuggestionDetailsOnSelect` API]
+  (https://github.com/atom/autocomplete-plus/wiki/Provider-API#defining-a-provider)
+Values can be either of:
+- `MethodCompletionInfo`: keeps (temporary) information of this `REPLCompletions.MethodCompletion`
+  + `f`: a function object
+  + `m`: a `Method` object
+  + `tt`: an input `Tuple` type
+- `MethodCompletionDetail`: stores the lazily computed detail of this `REPLCompletions.MethodCompletion`
+  + `rt`: `String` representing a return type of this method call
+  + `desc`: `String` representing documentation of this method
+
+If the second element is still a `MethodCompletionInfo` instance, that means its details
+  aren't lazily computed yet.
+If the second element is already a `MethodCompletionDetail` then we can reuse this cache.
+
+!!! note
+    If a method gets updated (i.e. redefined etc), key hash for the `REPLCompletions.MethodCompletion`
+      will be different from the previous one, so it's okay if we just rely on the hash key.
+
+!!! warning
+    Within the current implementation, we can't reflect changes that happens in backedges.
```

Comment:
Maybe the possible problem of this approach is this:
E.g. when we defined
```julia
f() = nothing
g() = f()
```
and invoke method completion for `g` (and cache the inference result `Nothing`),
and after that change `f` like
```julia
f() = 1
```
then, the rt still shows `Nothing`.

I don't think we need to be so strict on this (as in Julia's internal) though.

Reply:
I think that's fine for now. I feel like it should be possible to fix this, but we can always do that later.